### ISSUE TITLE
Converting between Deeprobust and pytorch geometric

### ISSUE THREAD

**Issue by akul-goyal** [DSE-MSU/DeepRobust#118]
Hi,

I am trying to attack my own custom model using the topology attack listed graph/global_attack. The attack runs fine until [here](https://github.com/DSE-MSU/DeepRobust/blob/a03ee5fb79df922183d821cdeeaed545c6b91aee/deeprobust/graph/global_attack/topology_attack.py#L112) where I have a modified adj matrix that I am looking to pass into my pytorch geometric model. Given it is in an adj matrix format rather than the edge_index format that pytorch geometric uses, how can I convert my adj_matrix to a format that can be used by pytorch without losing any of the gradients that are later needed for the backprop on [line](https://github.com/DSE-MSU/DeepRobust/blob/a03ee5fb79df922183d821cdeeaed545c6b91aee/deeprobust/graph/global_attack/topology_attack.py#L115). I tried doing adj_matrix.nonzero() but that gets rid of the gradients.

**Comment by ChandlerBang** [comment]
Hi,

The current code does not support attacking GAT using the topology attack. You may refer to the [PRBCD attack](https://github.com/sigeisler/robustness_of_gnns_at_scale/blob/main/rgnn_at_scale/attacks/prbcd.py#L128) to implement the attack with PyG.

However, we will soon include PRBCD in deeprobust based on PyG (maybe in one month). Please stay tuned :)

**Comment by akul-goyal** [comment]
Hi @ChandlerBang,

Thanks for the quick response. Is it possible to provide some intuition on how PRBCD is attacking PYG? In the current format of deep robust, a dense matrix gets modified. When I convert between deep robust to PYG the gradients of the dense matrix are lost. How can you preserve this? Furthermore, I am using a [neighborsampler](https://pytorch-geometric.readthedocs.io/en/latest/_modules/torch_geometric/loader/neighbor_sampler.html) for pyg so is there a way to pass in the dense matrix to the neighbor sampler without losing the gradients?

**Comment by ChandlerBang** [comment]
Hey, you may take a look at [their paper](https://arxiv.org/abs/2110.14038) first. Basically, at each step they will sample a block of edge indices (a small portion of all the possible edge indices) and optimize the edge weights by gradient descent.

**Comment by YaningJia** [comment]
@akul-goyal Hi
I meet the same quetions. I'm trying to implement topology on gat, first, i use pyg to do it and use the same means as you, then to avoid the probleam of adj, i use pytorch, but that both get rid of the gradients.  do you solve it?

**Comment by akul-goyal** [comment]
Yes, I used the following for help! https://github.com/pyg-team/pytorch_geometric/issues/1511

**Comment by YaningJia** [comment]
Thanks a lot

**Comment by YaningJia** [comment]
@akul-goyal Hi
I'm sorry to bother you again. According to the link you provided, I do following modifications:
edge_index = adj.nonzero().t()
row, col = edge_index
edge_weight = adj[row, col]

and it solves the problems of vanishing gradients. And I use modified_adj as input as gat.  But the test accuracy of modified_adj is not apprent decline. I wonder if your topology attack is valid. And can you show me some codes about your modification if available? Thank you.

**Comment by akul-goyal** [comment]
Hey, I am not sure if I understand you correctly, but based on your code, nonzero() does not preserve gradients. So that may be the problem,

**Comment by YaningJia** [comment]
The aims of nonzeros is convert dense adj to edge_index which is suit for pyg. So if i use pyg, what should i do?
